com.unity.multiplayer.mlapi/Tests/Editor/InternalMessageHandlerTests.cs

```diff
@@ -0,0 +1,387 @@
+using System;
+using System.Collections.Generic;
+using System.IO;
+using MLAPI.Configuration;
+using MLAPI.Internal;
+using MLAPI.Messaging;
+using MLAPI.Messaging.Buffering;
+using MLAPI.Serialization;
+using MLAPI.Transports;
+using MLAPI.Transports.Tasks;
+using NUnit.Framework;
+using UnityEngine;
+using UnityEngine.SceneManagement;
+using UnityEngine.TestTools;
+using Object = UnityEngine.Object;
+
+namespace MLAPI.EditorTests
+{
+    public class InternalMessageHandlerTests
+    {
+        [Test]
+        public void MessageHandlerReceivedMessageServerClient()
+        {
+            // Init
+            var gameObject = new GameObject(nameof(MessageHandlerReceivedMessageServerClient));
+            var networkManager = gameObject.AddComponent<NetworkManager>();
+            var transport = gameObject.AddComponent<DummyTransport>();
+
+            // MLAPI sets this in validate
+            networkManager.NetworkConfig = new NetworkConfig()
+            {
+                // Set the current scene to prevent unexpected log messages which would trigger a failure
+                RegisteredScenes = new List<string>() {SceneManager.GetActiveScene().name}
+            };
+
+            // Set dummy transport that does nothing
+            networkManager.NetworkConfig.NetworkTransport = transport;
+
+            // Replace the real message handler with a dummy one that just prints a result
+            networkManager.MessageHandler = new DummyMessageHandler();
+
+            using (var inputBuffer = new NetworkBuffer())
+            {
+                // Start server since pre-message-handler passes IsServer & IsClient checks
+                networkManager.StartServer();
+
+                // Disable batching to make the RPCs come straight through
+                // This has to be done post start
+                networkManager.RpcQueueContainer.EnableBatchedRpcs(false);
+
+                // Should cause log (server only)
+                LogAssert.Expect(LogType.Log, nameof(MessageHandlerReceivedMessageServerClient) + " " + nameof(DummyMessageHandler.HandleConnectionRequest));
+                using (var messageStream = MessagePacker.WrapMessage(NetworkConstants.CONNECTION_REQUEST, inputBuffer))
```

Comment:
I'd encourage each one of these to be a separate test case. Throwing them all into one test means that if one fails, none of the following ones even get tested. Also makes it harder to maintain the test coverage.